_posts/2021-09-03-rudiments-of-data-wrangling-in-rust.md

```diff
@@ -0,0 +1,514 @@
+
+    let count = registrations.entry(year).or_insert(0);
+    *count += 1;
+
+The first statement gets a reference to the count, and the second
+statement dereferences it, to increase its value by one. You can
+read the first statement as "give me the entry, or else insert a
+zero as its value".
+
+Note that in order to do this, the hashmap needs to be declared
+as mutable, like I did:
+
+    let mut registrations = HashMap::<i32, u32>::new();
+
+## Print out the results
+
+Now that I've iterated through all the records in the CSV file,
+I should have the first registration counts of all electric
+cars from the years 2016 to 2020 inclusive in the hashmap,
+so I can now proceed to print them in the console.
+
+Rust uses the `println!` macro for printing. The exclamation
+mark tells you without a doubt that this is a macro invocation.
+Rust macros reduce boilerplate code in many
+common situations. For example, initializing a vector with
+the `vec!` macro lets you focus on the initial contents instead
+of the code that allocates memory and adds each individual 
+element.
+
+The `println!` macro also takes care of formatting the data.
+You can let Rust print a value like it pleases, but you can
+also utilize the comprehensive formatting templates (also used
+by the `format!` macro).
+
+Since I stored the range of the interesting years earlier,
+I can use it when I present the results. First I print out a
+heading:
+
+    println!("EV registrations {}-{} by year:", 
+        year_range.start(), year_range.end());
+
+Then, because I'm confident that I have an entry in the hashmap
+for each of the years (remember, if there wasn't already an entry,
+I made one, and initialized it to zero), I can just loop through
+the range and print out the result for each year.
+
+    for year in year_range {
+        println!("{:>5}: {:>5}", 
+            year, registrations.get(&year).unwrap());
+    }
+
+Getting an entry from the hashmap can fail, but in this case 
+I just use the common Rust technique of "unwrapping" the result.
+Keep in mind that if you're not writing a quick and dirty utility
+like this one, unwrapping is not the best thing to do, because
+failure will cause the program to panic. Using unwrap is somewhat
+opinion based, but in a case like this I think it can be justified.
+
+Finally, I can print the total number.
+
+    println!("Total: {:>5}", electric_car_count);
+
+Just by eyeballing the numbers you can see that the growth 
+in EV registrations has been quite rapid. I could have also 
+computed the count while printing out the results, instead
+of collecting it as we progressed through the records.
+
+## There's plenty more where that came from
+
+This has been a whirlwind tour of some common real-world tasks
+involved in writing a Rust program. However, even in this small
+program there are many Rust features that I was not able to
+explain in depth, such as borrowing. (Even lifetimes already made a 
```

Comment:
in depth -> in-depth

Reply:
Intentional, see e.g. https://ell.stackexchange.com/questions/102701/what-is-the-difference-between-in-depth-and-in-depth.

So basically, "to make an in-depth analysis", but "to analyse in depth".